x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/StreamingResponseHandler.java

```diff
@@ -0,0 +1,156 @@
+        this.responseHandler = responseHandler;
+        this.listener = listener;
+    }
+
+    @Override
+    public void subscribe(Flow.Subscriber<? super HttpResult> subscriber) {
+        if (downstream != null) {
+            subscriber.onError(
+                new IllegalStateException("Failed to initialize streaming response. Another subscriber is already subscribed.")
+            );
+            return;
+        }
+
+        downstream = subscriber;
+        subscriber.onSubscribe(forwardingSubscription());
+    }
+
+    private Flow.Subscription forwardingSubscription() {
+        return new Flow.Subscription() {
+            @Override
+            public void request(long n) {
+                if (upstreamIsClosed.get()) {
+                    downstream.onComplete(); // shouldn't happen, but reinforce that we're no longer listening
+                } else if (upstream != null) {
+                    upstream.request(n);
+                } else {
+                    // this shouldn't happen, the expected call pattern is onNext -> subscribe after the listener is invoked
+                    var errorMessage = "Failed to initialize streaming response. onSubscribe must be called first to set the upstream";
+                    assert false : errorMessage;
+                    downstream.onError(new IllegalStateException(errorMessage));
+                }
+            }
+
+            @Override
+            public void cancel() {
+                if (upstreamIsClosed.compareAndSet(false, true) && upstream != null) {
+                    upstream.cancel();
+                }
+            }
+        };
+    }
+
+    @Override
+    public void onSubscribe(Flow.Subscription subscription) {
+        upstream = subscription;
+        // start the first request, which will call onNext and validate the first HttpResult
+        upstream.request(1);
+    }
+
+    @Override
+    public void onNext(HttpResult item) {
+        if (processedFirstItem.compareAndSet(false, true)) {
+            try {
+                responseHandler.validateResponse(throttlerManager, throttlerLogger, request, item);
+                var inferenceServiceResults = responseHandler.parseResult(request, item, this);
+                assert downstream != null : "the responseHandler must invoke the subscribe method";
+                listener.onResponse(inferenceServiceResults);
+            } catch (Exception e) {
+                logException(throttlerLogger, request, item, responseHandler.getRequestType(), e);
+                listener.onFailure(e);
+                upstream.cancel();
+                onError(e);
+            }
+        } else {
+            downstream.onNext(item);
+        }
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        if (upstreamIsClosed.compareAndSet(false, true)) {
+            if (downstream != null) {
+                downstream.onError(throwable);
+            } else {
+                log.debug(
```

Comment:
```suggestion
                log.warn(
```

Looks like something unexpected happened

Reply:
Do we have to use the `throttleManager` for that?